File: slp/backend-dotnet/Features/Dashboard/DbWordOfTheDayProvider.cs
using backend_dotnet.Data;
using Microsoft.EntityFrameworkCore;

namespace backend_dotnet.Features.Dashboard;

public class DbWordOfTheDayProvider : IWordOfTheDayProvider
{
    private readonly AppDbContext _db;

    public DbWordOfTheDayProvider(AppDbContext db)
    {
        _db = db;
    }

    public async Task<WordOfTheDayDto> GetWordOfTheDayAsync()
    {
        var today = DateTime.UtcNow.Date;

        // Try to get today's word
        var wordEntity = await _db.DailyWords
            .FirstOrDefaultAsync(w => w.TargetDate == today);

        // If none exists for today, fallback to the most recent past word
        if (wordEntity == null)
        {
            wordEntity = await _db.DailyWords
                .Where(w => w.TargetDate <= today)
                .OrderByDescending(w => w.TargetDate)
                .FirstOrDefaultAsync();
        }

        // If still null (table empty), return a friendly placeholder
        if (wordEntity == null)
        {
            return new WordOfTheDayDto
            {
                Word = "Welcome!",
                PartOfSpeech = "",
                VietnameseTranslation = "Chào mừng",
                Example = "No word of the day yet. Please add some entries to the daily_word table.",
                Origin = null,
                FunFact = null
            };
        }

        return new WordOfTheDayDto
        {
            Word = wordEntity.Word,
            PartOfSpeech = wordEntity.PartOfSpeech ?? string.Empty,
            VietnameseTranslation = wordEntity.VietnameseTranslation ?? string.Empty,
            Example = wordEntity.Example ?? string.Empty,
            Origin = wordEntity.Origin,
            FunFact = wordEntity.FunFact
        };
    }
}

File: slp/backend-nestjs/src/modules/dashboard/providers/db-word-of-the-day.provider.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, LessThanOrEqual } from 'typeorm';
import { IWordOfTheDayProvider } from '../interfaces/word-of-the-day-provider.interface';
import { WordOfTheDayDto } from '../dto/word-of-the-day.dto';
import { DailyWord } from '../daily-word.entity';

@Injectable()
export class DbWordOfTheDayProvider implements IWordOfTheDayProvider {
  constructor(
    @InjectRepository(DailyWord)
    private readonly dailyWordRepo: Repository<DailyWord>,
  ) {}

  async getWordOfTheDayAsync(): Promise<WordOfTheDayDto> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    // Try today's word
    let wordEntity = await this.dailyWordRepo.findOne({
      where: { targetDate: today },
    });

    // Fallback to most recent past word
    if (!wordEntity) {
      wordEntity = await this.dailyWordRepo.findOne({
        where: { targetDate: LessThanOrEqual(today) },
        order: { targetDate: 'DESC' },
      });
    }

    // Fallback placeholder
    if (!wordEntity) {
      return {
        word: 'Welcome!',
        partOfSpeech: '',
        vietnameseTranslation: 'Chào mừng',
        example: 'No word of the day yet. Please add some entries to the daily_word table.',
        origin: undefined,
        funFact: undefined,
      };
    }

    return {
      word: wordEntity.word,
      partOfSpeech: wordEntity.partOfSpeech ?? '',
      vietnameseTranslation: wordEntity.vietnameseTranslation ?? '',
      example: wordEntity.example ?? '',
      origin: wordEntity.origin,
      funFact: wordEntity.funFact,
    };
  }
}